src/main/java/org/apache/mesos/hdfs/state/ClusterState.java

```diff
@@ -70,57 +72,35 @@ public State getState() {
   }
 
   public boolean notInDfsHosts(String host) {
-    return !dfsHosts.contains(host);
+    return !taskSlaveMap.values().contains(host);
   }
 
-  public void addTask(Protos.TaskID taskId, Scheduler.DfsTask dfsTask) {
-    tasks.put(taskId, dfsTask);
-    Scheduler.DfsTask.Type type = dfsTask.type;
-    switch (type) {
-      case NN :
-        nameNodeHosts.add(dfsTask.hostname);
-        break;
-      case JN :
-        journalNodeHosts.add(dfsTask.hostname);
-        break;
-      default :
-        break;
+  public void addTask(Protos.TaskID taskId, String hostname, String slaveId) {
+    taskHostMap.put(taskId, hostname);
+    taskSlaveMap.put(taskId, slaveId);
+    if (taskId.getValue().contains(HDFSConstants.NAME_NODE_TASKID)) {
+      nameNodeHosts.add(hostname);
+    } else if (taskId.getValue().contains(HDFSConstants.JOURNAL_NODE_ID)) {
+      journalNodeHosts.add(hostname);
     }
   }
 
   public void updateTask(Protos.TaskStatus taskStatus) {
-    if (tasks.containsKey(taskStatus.getTaskId())) {
-      tasks.get(taskStatus.getTaskId()).taskStatus = taskStatus;
-      Scheduler.DfsTask.Type type = tasks.get(taskStatus.getTaskId()).type;
-      dfsHosts.add(taskStatus.getSlaveId().getValue());
-      switch (type) {
-        case DN :
-        case ZKFC :
-          break;
-        case NN :
-          nameNodes.add(taskStatus.getTaskId());
-          break;
-        case JN :
-          journalNodes.add(taskStatus.getTaskId());
-          break;
-        default :
-          Scheduler.log.error("Unknown task type: " + type);
-          break;
-      }
-    } else {
-      log.error("Asked to update unknown task, taskId=" + taskStatus.getTaskId().getValue());
+    if (taskStatus.getTaskId().getValue().contains(HDFSConstants.NAME_NODE_TASKID)) {
+      nameNodes.add(taskStatus.getTaskId());
+    } else if (taskStatus.getTaskId().getValue()
+        .contains(HDFSConstants.JOURNAL_NODE_ID)) {
+      journalNodes.add(taskStatus.getTaskId());
     }
   }
 
   public void removeTask(Protos.TaskStatus taskStatus) {
     Protos.TaskID taskId = taskStatus.getTaskId();
-    if (tasks.containsKey(taskId) && tasks.get(taskId).hostname != null) {
-      journalNodeHosts.remove(tasks.get(taskId).hostname);
-      nameNodeHosts.remove(tasks.get(taskId).hostname);
-    }
-    tasks.remove(taskId);
+    journalNodeHosts.remove(taskHostMap.get(taskId));
+    nameNodeHosts.remove(taskHostMap.get(taskId));
+    taskHostMap.remove(taskId);
     nameNodes.remove(taskId);
     journalNodes.remove(taskId);
-    dfsHosts.remove(taskStatus.getSlaveId().getValue());
+    taskSlaveMap.remove(taskId);
```

Comment:
In the state-machine version, we'll only update the type of task that was actually failed/removed.
Either that or we'll need to kill the JN,NN,ZKFC if we're removing them.
Or should removing them just ensure that we'll try to relaunch them, which will cleanly (idempotently) take care of it?